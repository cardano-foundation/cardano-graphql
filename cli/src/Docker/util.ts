import { QuestionCollection } from 'inquirer'
import { DockerComposeStack, StatefulService } from './index'

export function selectServicesFromStack (stack: DockerComposeStack, multi = true): QuestionCollection {
  const label = multi === true ? 'services' : 'service'
  return {
    name: label,
    message: `Select ${label}`,
    type: multi === true ? 'checkbox' : 'list',
    choices: [...stack.statefulServices.values()].map(service => {
      return {
        name: service.name,
        value: service
      }
    }),
    validate (input) {
      if (input.length === 0) {
        throw new Error('At least one service must be selected')
      }
      return true
    }
  }
}

export function servicesFromNameArg (stack: DockerComposeStack, names: string): StatefulService[] {
  return names.split(',').map((name: string) => {
    if (!stack.statefulServices.has(name)) {
      throw new Error(`Unknown service ${name}`)
    }
    return stack.statefulServices.get(name)
  })
}
