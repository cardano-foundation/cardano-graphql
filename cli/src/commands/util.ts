import { DockerComposeStack } from '../docker'

export function serviceChoices (stack: DockerComposeStack) {
  return [...stack.statefulServices.values()].map(service => {
    return { name: service.name, value: service }
  })
}

export function serviceFromName (stack: DockerComposeStack, name: string) {
  if (!stack.statefulServices.has(name)) {
    throw new Error(`Unknown service ${name}`)
  }
  return stack.statefulServices.get(name)
}

export function servicesFromNames (stack: DockerComposeStack, names: string) {
  return names.split(',').map((name: string) => serviceFromName(stack, name))
}
