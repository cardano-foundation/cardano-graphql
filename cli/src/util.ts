
export function ensureValue (input: string) {
  return !input ? 'Cannot be blank' : true
}
