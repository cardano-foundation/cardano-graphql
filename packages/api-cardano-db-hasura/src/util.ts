import { ConnectionConfig, createInteractionContext } from '@cardano-ogmios/client'
import { Logger } from 'ts-log'

export const createInteractionContextWithLogger = (connection: ConnectionConfig, logger: Logger, module: string, onClose?: () => Promise<void>) =>
  createInteractionContext(
    (error) => {
      logger.error({ module, error }, error.message)
    },
    async (code, reason) => {
      if (code === 1000) {
        logger.info({ module, code }, reason)
      } else {
        logger.error({ module, code }, 'Connection closed')
        await onClose?.()
      }
    },
    {
      connection
    }
  )
