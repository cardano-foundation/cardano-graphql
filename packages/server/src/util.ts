import { Application } from 'express'
import http from 'http'

export function listenPromise (app: Application, port: number, hostname?: string): Promise<http.Server> {
  return new Promise(function (resolve, reject) {
    const server: http.Server = app.listen(port, hostname, () => resolve(server))
    server.on('error', reject)
  })
}
