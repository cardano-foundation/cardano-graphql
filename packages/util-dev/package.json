{
  "name": "@cardano-graphql/util-dev",
  "version": "1.0.0",
  "description": "Common development utilities",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -b ./src",
    "cleanup": "shx rm -rf dist node_modules",
    "lint": "eslint --ignore-path ../../.eslintignore \"**/*.ts\"",
    "prepack": "yarn build",
    "test": "shx echo No tests in this package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/input-output-hk/cardano-graphql.git"
  },
  "author": "Rhys Bartels-Waller",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/input-output-hk/cardano-graphql/issues"
  },
  "homepage": "https://github.com/input-output-hk/cardano-graphql/blob/master/packages/util-dev/README.md",
  "devDependencies": {
    "@cardano-graphql/util": "1.0.0",
    "shx": "^0.3.2"
  },
  "directories": {
    "lib": "src",
    "test": "test"
  },
  "files": [
    "dist",
    "LICENSE",
    "NOTICE",
    "README.md"
  ]
}
