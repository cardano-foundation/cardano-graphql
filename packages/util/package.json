{
  "name": "@cardano-graphql/util",
  "version": "1.0.0",
  "description": "Common utilities",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -b ./src",
    "cleanup": "shx rm -rf dist node_modules",
    "lint": "eslint --ignore-path ../../.eslintignore \"**/*.ts\"",
    "prepack": "yarn build",
    "test": "shx echo No tests in this package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/input-output-hk/cardano-graphql.git"
  },
  "author": "Rhys Bartels-Waller",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/input-output-hk/cardano-graphql/issues"
  },
  "homepage": "https://github.com/input-output-hk/cardano-graphql/blob/master/packages/util/README.md",
  "dependencies": {
    "graphql": "14.5.8",
    "graphql-bigint": "^1.0.0",
    "graphql-tag": "^2.10.3",
    "p-retry": "^4.2.0"
  },
  "devDependencies": {
    "@types/node": "^14.0.13",
    "shx": "^0.3.2",
    "typescript": "^3.9.5"
  },
  "directories": {
    "lib": "src",
    "test": "test"
  },
  "files": [
    "dist",
    "LICENSE",
    "NOTICE",
    "README.md"
  ]
}
