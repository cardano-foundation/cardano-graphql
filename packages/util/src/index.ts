import { onFailedAttemptFor } from './onFailedAttemptFor'
import { loadQueryNode } from './queryNodeLoading'
import * as dateTimeValidation from './dateTimeValidation'
import * as scalars from './scalars'

export default {
  dateTimeValidation,
  onFailedAttemptFor,
  loadQueryNode,
  scalars
}
