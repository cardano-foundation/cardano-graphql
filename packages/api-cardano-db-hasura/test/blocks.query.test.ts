import path from 'path'
import { DocumentNode } from 'graphql'
import gql from 'graphql-tag'
import util from '@cardano-graphql/util'
import { TestClient } from '@cardano-graphql/util-dev'
import { block29021, block29022 } from './data_assertions'
import { buildClient } from './util'

function loadQueryNode (name: string): Promise<DocumentNode> {
  return util.loadQueryNode(path.resolve(__dirname, '..', 'src', 'example_queries', 'blocks'), name)
}

describe('blocks', () => {
  let client: TestClient
  beforeAll(async () => {
    client = await buildClient()
  }, 60000)

  it('caps the response to 100 blocks', async () => {
    const result = await client.query({
      query: await loadQueryNode('blockHashesNoArgs')
    })
    expect(result.data.blocks.length).toBe(100)
  })

  it('allows custom pagination size with a limit and offset', async () => {
    const page1 = await client.query({
      query: await loadQueryNode('first20Blocks')
    })
    const page2 = await client.query({
      query: await loadQueryNode('second20Blocks')
    })
    expect(page1.data.blocks.length).toBe(20)
    expect(page1.data.blocks[19].number).toBe(23)
    expect(page2.data.blocks.length).toBe(20)
    expect(page2.data.blocks[19].number).toBe(43)
  })

  it('Can return blocks by number', async () => {
    const result = await client.query({
      query: await loadQueryNode('blockByNumber'),
      variables: { number: 29022 }
    })
    expect(result.data.blocks.length).toBe(1)
    expect(result.data.blocks[0]).toEqual({ hash: block29022.basic.hash })
    expect(result.data).toMatchSnapshot()
  })

  // Todo: Restore. Assertion is valid, but matcher seems to have 'ArrayContaining [ObjectContaining' appended
  // it('Can return blocks by an array of hashes', async () => {
  //   const result = await client.query({
  //     query: await loadQueryNode('blocksByHashes'),
  //     variables: { hashes: [block29021.basic.hash, block29022.basic.hash] }
  //   })
  //   expect(result.data.blocks.length).toBe(2)
  //   console.log()
  //   expect(result.data.blocks).toEqual(
  //     expect.arrayContaining([
  //       expect.objectContaining(block29021.basic),
  //       expect.objectContaining(block29022.basic)
  //     ])
  //   )
  // })

  it('Can return aggregated data', async () => {
    const result = await client.query({
      query: await loadQueryNode('aggregateDataWithinBlock'),
      variables: { number: 29021, epochLessThan: 185 }
    })
    expect(result.data.blocks[0]).toEqual(block29021.aggregated)
    expect(result.data).toMatchSnapshot()
  })

  it('Can return filtered aggregated data', async () => {
    const result = await client.query({
      query: gql`query {
          blocks( where: { number: { _eq: 29021 }}) {
              transactions_aggregate(
                  where: {
                      _and: [
                          { fee: { _gt: "10" }},
                          { totalOutput: { _lt: "4924799478670" } }
                      ]
                  }) {
                  aggregate {
                      count
                  }
              }
              number
          }
      }`
    })
    expect(result.data.blocks[0]).toEqual(block29021.aggregated_filtered)
    expect(result.data).toMatchSnapshot()
  })

  it('are linked to their predecessor, and the chain can be traversed', async () => {
    const result = await client.query({
      query: await loadQueryNode('selectGreatGrandparentBlock'),
      variables: { number: 29022 }
    })
    expect(result.data.blocks[0].previousBlock.previousBlock.previousBlock.number).toBe(29019)
    expect(result.data).toMatchSnapshot()
  })

  it('are linked to their successor, and the chain can be traversed', async () => {
    const result = await client.query({
      query: await loadQueryNode('selectGreatGrandchildBlock'),
      variables: { number: 29022 }
    })
    expect(result.data.blocks[0].nextBlock.nextBlock.nextBlock.number).toBe(29025)
    expect(result.data).toMatchSnapshot()
  })
})
