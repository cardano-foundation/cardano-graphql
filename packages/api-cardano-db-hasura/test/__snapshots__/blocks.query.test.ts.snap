// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blocks Can return aggregated data 1`] = `
Object {
  "blocks": Array [
    Object {
      "number": 29021,
      "transactions_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "fee": 171158,
            "size": 218,
            "totalOutput": 2846601153795,
          },
          "count": "2",
          "max": Object {
            "fee": "171246",
            "size": "220",
            "totalOutput": "4924799478660",
          },
          "min": Object {
            "fee": "171070",
            "size": "216",
            "totalOutput": "768402828930",
          },
          "sum": Object {
            "deposit": "0",
            "fee": "342316",
            "size": "436",
            "totalOutput": "5693202307590",
          },
        },
      },
    },
  ],
}
`;

exports[`blocks Can return blocks by number 1`] = `
Object {
  "blocks": Array [
    Object {
      "hash": "bf13d9a80ad99a4f34edb8a3262dd8120e29bbe182732cd3b00bf3d1bb7c2380",
    },
  ],
}
`;

exports[`blocks Can return filtered aggregated data 1`] = `
Object {
  "blocks": Array [
    Object {
      "number": 29021,
      "transactions_aggregate": Object {
        "aggregate": Object {
          "count": "2",
        },
      },
    },
  ],
}
`;

exports[`blocks are linked to their predecessor, and the chain can be traversed 1`] = `
Object {
  "blocks": Array [
    Object {
      "hash": "bf13d9a80ad99a4f34edb8a3262dd8120e29bbe182732cd3b00bf3d1bb7c2380",
      "previousBlock": Object {
        "number": 29021,
        "previousBlock": Object {
          "number": 29020,
          "previousBlock": Object {
            "number": 29019,
          },
        },
      },
    },
  ],
}
`;

exports[`blocks are linked to their successor, and the chain can be traversed 1`] = `
Object {
  "blocks": Array [
    Object {
      "hash": "bf13d9a80ad99a4f34edb8a3262dd8120e29bbe182732cd3b00bf3d1bb7c2380",
      "nextBlock": Object {
        "nextBlock": Object {
          "nextBlock": Object {
            "number": 29025,
          },
          "number": 29024,
        },
        "number": 29023,
      },
    },
  ],
}
`;
